libsolv-sys/src/queue.rs

```diff
@@ -12,48 +13,93 @@ pub struct Queue {
 extern "C" {
     pub fn queue_init(q: *mut Queue);
     pub fn queue_init_buffer(q: *mut Queue, buf: *mut Id, size: c_int);
-    pub fn queue_init_clone(t: *mut Queue, s: *mut Queue); // FIXME: source is const?
+    pub fn queue_init_clone(t: *mut Queue, s: *mut Queue); // FIXME: source is const? - Yes
     pub fn queue_free(q: *mut Queue);
 
+    pub fn queue_alloc_one(q: *mut Queue); /* internal */
+    pub fn queue_alloc_one_head(q: *mut Queue); /* internal */
+
     pub fn queue_insert(q: *mut Queue, pos: c_int, id: Id);
     pub fn queue_insert2(q: *mut Queue, pos: c_int, id1: Id, id2: Id);
-    pub fn queue_insertn(q: *mut Queue, pos: c_int, n: c_int, elements: *mut Id); // FIXME: elements is const?
+    pub fn queue_insertn(q: *mut Queue, pos: c_int, n: c_int, elements: *mut Id); // FIXME: elements is const? - Yes.
     pub fn queue_delete(q: *mut Queue, pos: c_int);
     pub fn queue_delete2(q: *mut Queue, pos: c_int);
     pub fn queue_deleten(q: *mut Queue, pos: c_int, n: c_int);
     pub fn queue_prealloc(q: *mut Queue, n: c_int);
 
-    fn queue_empty_real(q: *mut Queue);
-    fn queue_shift_real(q: *mut Queue) -> Id;
-    fn queue_pop_real(q: *mut Queue) -> Id;
-    fn queue_unshift_real(q: *mut Queue, id: Id);
-    fn queue_push_real(q: *mut Queue, id: Id);
-    fn queue_pushunique_real(q: *mut Queue, id: Id);
-    fn queue_push2_real(q: *mut Queue, id1: Id, id2: Id);
-    fn queue_truncate_real(q: *mut Queue, n: c_int);
 }
 
-pub fn queue_empty(q: *mut Queue) {
-    unsafe { queue_empty_real(q) }
+pub unsafe fn queue_empty(q: *mut Queue) {
+    let ref mut queue = *q;
+    if !queue.alloc.is_null() {
+        queue.left += (queue.elements as usize - queue.alloc as usize) as c_int + queue.count;
```

Comment:
Because Rust does not allow raw pointer arithmetic. You can use the offset function to do some of the necessary arithmetic, but you can't subtract one pointer from another.